cmd/zdb/zdb.c

```diff
@@ -4187,6 +4196,264 @@ same_metaslab(spa_t *spa, uint64_t vdev, uint64_t off1, uint64_t off2)
 	return ((off1 >> ms_shift) == (off2 >> ms_shift));
 }
 
+/*
+ * Prints the elements in each row properly for -P output.
+ */
+static inline void
+prtrow_nicenum(uint64_t number, int width)
+{
+	char numbuf[32];
+
+	zdb_nicenum(number, numbuf, sizeof (numbuf));
+	if (dump_opt['P'])
+		(void) printf("\t%s", numbuf);
+	else
+		printf("%*s", width, numbuf);
+}
+
+/*
+ * Used to simplify reporting of the histogram data.
+ */
+typedef struct one_histo {
+	char *name;
+	uint64_t *count;
+	uint64_t *len;
+	uint64_t cumulative;
+} one_histo_t;
+
+/*
+ * The number of separate histograms processed for psize, lsize and asize.
+ */
+#define	NUM_HISTO 3
+
+/*
+ * This routine will create a fixed column size output of three different
+ * histograms showing:
+ * by blocksize of 0 - 2^ SPA_MAXBLOCKSHIFT the count, length and
+ * cumulative length of the psize, lsize and asize blocks.
+ *
+ * All three types of blocks are listed on a single line
+ *
+ * By default the table is printed in nicenumber format (e.g. 123K) but
+ * if the '-P' parameter is specified then the full raw number is printed
+ * out.
+ */
+static void
+dump_size_histograms(zdb_cb_t *zcb)
+{
+	/*
+	 * Keep track of the smallest and largest blocksize used in
+	 * this file system so that we don't print out the block
+	 * sizes that have zero count of blocks used.
```

Comment:
I'd suggest we always print all the rows to keep it easy to parse.  Since the binning is by powers of two there are only 16 rows of data.